src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -8,16 +8,128 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Diagnostics
 {
+    internal sealed class AsyncQueue<TElement>
+    {
+        private readonly Channel<TElement> _channel = Channel.CreateUnbounded<TElement>();
+        private readonly TaskCompletionSource<bool> _whenCompleted = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        /// <summary>
+        /// Gets a task that transitions to a completed state when <see cref="Complete"/> or
+        /// <see cref="TryComplete"/> is called.  This transition will not happen synchronously.
+        /// 
+        /// This Task will not complete until it has completed all existing values returned
+        /// from <see cref="DequeueAsync"/>.
+        /// </summary>
+        public Task WhenCompletedTask
+            => _whenCompleted.Task;
+
+        /// <summary>
+        /// Gets a value indicating whether the queue has completed.
+        /// </summary>
+        public bool IsCompleted
+            => _whenCompleted.Task.IsCompleted;
+
+        /// <summary>
+        /// The number of unconsumed elements in the queue.
+        /// </summary>
+        public int Count => _channel.Reader.Count;
+
+        /// <summary>
+        /// Adds an element to the tail of the queue.  This method will throw if the queue 
+        /// is completed.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is already completed.</exception>
+        /// <param name="value">The value to add.</param>
+        public void Enqueue(TElement value)
+        {
+            if (!TryEnqueue(value))
+                throw new InvalidOperationException($"Cannot call {nameof(Enqueue)} when the queue is already completed.");
+        }
+
+        /// <summary>
+        /// Tries to add an element to the tail of the queue.  This method will return false if the queue
+        /// is completed.
+        /// </summary>
+        /// <param name="value">The value to add.</param>
+        public bool TryEnqueue(TElement value)
+            => _channel.Writer.TryWrite(value);
+
+        /// <summary>
+        /// Attempts to dequeue an existing item and return whether or not it was available.
+        /// </summary>
+        public bool TryDequeue(out TElement d)
+            => _channel.Reader.TryRead(out d);
```

Comment:
Count/Enqueue/TryEnqueue/TryDequeue all seem sane, just using the same semantics as channel.